docs/architecture/grpc-for-wcf-developers/types-of-rpc.md

```diff
@@ -7,5 +7,258 @@ ms.date: 09/02/2019
+
+## WCF duplex, one-way to client
+
+WCF applications (with certain bindings) can create a persistent connection between client and server, and the server can asynchronously send data to the client until the connection is closed, using a *callback interface*.
+
+gRPC services can work with streams of messages, in either or both directions. This does not map exactly to WCF Duplex, but the same result can be achieved.
+
+Here is an example of a server streaming RPC.
+
+```protobuf
+service ClockStreamer {
+    rpc Subscribe(ClockSubscribeRequest) returns (stream ClockMessage);
+}
+```
+
+```csharp
+public class ClockStreamerService : ClockStreamer.ClockStreamerBase
+{
+    public override async Task Subscribe(ClockSubscribeRequest request,
+        IServerStreamWriter<ClockMessage> responseStream,
+        ServerCallContext context)
+    {
+        while (!context.CancellationToken.IsCancellationRequested)
+        {
+            var time = DateTimeOffset.UtcNow;
+            await responseStream.WriteAsync(new ClockMessage { message = $"The time is {time:t}." });
+            await Task.Delay(TimeSpan.FromSeconds(10), context.CancellationToken);
+        }
+    }
+}
+```
+
+This server stream could be consumed from a client application as shown here.
+
+```csharp
+public async Task TellTheTimeAsync(CancellationToken token)
+{
+    var channel = GrpcChannel.ForAddress("https://localhost:5001");
+    var client = new ClockStreamer.ClockStreamerClient(channel);
+
+    var request = new ClockSubscribeRequest();
+    var response = client.Subscribe(request);
+
+    await foreach (var update in response.ResponseStream.ReadAllAsync(token))
+    {
+        Console.WriteLine(update.Message);
+    }
+}
+```
+
+> [!NOTE]
+> Server streaming RPCs are useful for subscription-style services, and also for sending very large datasets when it would be inefficient or impossible to build the entire dataset in memory. However, streaming responses is not as fast as sending `repeated` fields in a single message, so as a rule streaming should not be used for small datasets.
+
+### Differences to WCF
+
+A WCF duplex service uses a client callback interface that can have multiple methods. A gRPC server streaming service can only send messages over a single stream. If you need multiple methods, use a message type with either [an Any field or a oneof field](protobuf-any-oneof.md) to send different messages, and write code in the client to handle them.
+
+In WCF, the one-way method called from the client returns immediately, but the `ServiceContract` class with the session is kept alive until the connection is terminated. In gRPC, the Task returned by the implementation method should not complete until the connection is closed.
```

Comment:
ServiceContract here refers to this class? https://docs.microsoft.com/en-us/dotnet/api/system.servicemodel.servicecontractattribute?view=netframework-4.8

Reply:
Yes

Reply:
Added link